cpp/src/dictionary/dictionary_factories.cu

```diff
@@ -0,0 +1,186 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/dictionary/dictionary_factories.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/gather.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+
+namespace cudf
+{
+namespace detail
+{
+namespace
+{
+
+// Sort functor for column row positions.
+template <bool has_nulls = true>
+struct sort_functor
+{
+    column_device_view d_column;
+    __host__ __device__ bool operator()(size_type lhs_index, size_type rhs_index)
+    {
+        auto comparator = experimental::element_relational_comparator<has_nulls>{
+                          d_column, d_column, null_order::AFTER }; // put nulls at the end
+        auto result = experimental::type_dispatcher(d_column.type(), comparator,
+                                                    lhs_index, rhs_index);
+        return result == experimental::weak_ordering::LESS; // always sort ascending
+    }
+};
+
+// Copy functor used for locating and marking unique values.
+template <bool has_nulls = true>
+struct copy_unique_functor
+{
+    column_device_view d_column;
+    const int32_t* d_ordinals;
+    int32_t* d_indices;
+
+    __device__ bool operator()(int32_t idx)
+    {
+        if( idx==0 )
+        {   // first element is always unique
+            d_indices[0] = 0;
+            return true;
+        }
+        // check if adjacent elements match
+        auto lhs_index = d_ordinals[idx-1];
+        auto rhs_index = d_ordinals[idx];
+        auto result = !cudf::experimental::type_dispatcher( d_column.type(),
+                             experimental::element_equality_comparator<has_nulls>{d_column, d_column, true},
+                             lhs_index, rhs_index);
+        d_indices[idx] = static_cast<int32_t>(result); // convert bool to integer [0,1]
+        return result;
+    }
+};
+
+} // namespace
+} // detail
+
+/**
+ * @brief Create dictionary column from a column_view.
+ *
+ */
+std::unique_ptr<column> make_dictionary_column( column_view const& input_column,
+                                                rmm::mr::device_memory_resource* mr,
+                                                cudaStream_t stream)
+{
```

Comment:
After staring at this for a while, this function could be simplified significantly and reuse existing logic by doing:
- `unique_keys = cudf::drop_duplicates(input_column)`
- `indices = cudf::lower_bound(unique_keys, input_column)`

Reply:
> After staring at this for a while, this function could be simplified significantly and reuse existing logic by doing:
> 
> * `unique_keys = cudf::drop_duplicates(input_column)`
> * `indices = cudf::lower_bound(unique_keys, input_column)`

So after looking at these two function they do not directly appear to achieve the result we want without an extra key sort. The `drop_duplicates` appears to remove duplicates and but _should_ not return sorted results.  I say _should_ because it appears the current implementation does return sorted results which I think is a defect. Therefore, an extra sort _would_ be required to create the keys once this is fixed. The `lower_bound` should give the right result for the indices but will incur unnecessary, expensive compare operations.

One goal of this function was to minimize the compares of the key values (which will likely be strings). The `drop_duplicate/lower_bound` approach would incur key compares for two sorts and a key compare for `copy_if` to remove duplicates followed by the key compares needed for `lower_bound`.
The current implementation requires only one key compare for `sort` and one key compare for `copy_if`. The rest of the kernels operate on integers.

Reply:
> I say should because it appears the current implementation does return sorted results which I think is a defect.

It's not a defect. It was decided that results should be returned in sorted order because it would just be unnecessary overhead to return results in a stable order. If the caller desired the original order, they can pass in a non-key column that simply indexes the rows `{0...n-1}` which can then be used as a gather map. (The documentation can/should definitely be improved to describe this intention.)

I think your concern about minimizing the number of key comparisons is definitely valid. In the spirit of DRY I'd really like to find a way to reuse as much existing logic as possible. That might mean modifying/adding internal `detail::` APIs for `drop_duplicates/lower_bound` that operate on iterators to make them more flexible for internal use like in this function. Could you explore if it would be possible to limit the number of key comparisons and still re-use existing logic?